src/Compilers/CSharp/Portable/Binder/Binder_TupleOperators.cs

```diff
@@ -0,0 +1,317 @@
+            var rightTupleType = MakeConvertedType(operatorsBuilder, node.Right, right, nullable, compilation, diagnostics, isRight: true);
+
+            return new TupleBinaryOperatorInfo.Nested(operatorsBuilder.ToImmutableAndFree(), leftTupleType, rightTupleType);
+        }
+
+        private static int? GetTupleCardinality(BoundExpression expr)
+        {
+            if (expr.Kind == BoundKind.TupleLiteral)
+            {
+                var tuple = (BoundTupleLiteral)expr;
+                return tuple.Arguments.Length;
+            }
+
+            var type = expr.Type;
+            if (type is null)
+            {
+                return null;
+            }
+
+            type = type.StrippedType();
+
+            if (type.IsTupleType)
+            {
+                return type.TupleElementTypes.Length;
+            }
+
+            return null;
+        }
+
+        private static ImmutableArray<BoundExpression> GetTupleArgumentsOrPlaceholders(BoundExpression expr)
+        {
+            if (expr.Kind == BoundKind.TupleLiteral)
+            {
+                return ((BoundTupleLiteral)expr).Arguments;
+            }
+
+            // placeholder bound nodes with the proper types are sufficient to bind the element-wise binary operators
+            return expr.Type.StrippedType().TupleElementTypes
+                .SelectAsArray(t => (BoundExpression)new BoundTupleOperandPlaceholder(expr.Syntax, t));
+        }
+
+        /// <summary>
+        /// Make a tuple type (with appropriate nesting) from the types (on the left or on the right) collected 
+        /// from binding element-wise binary operators.
+        /// </summary>
+        private TypeSymbol MakeConvertedType(ArrayBuilder<TupleBinaryOperatorInfo> operators, CSharpSyntaxNode syntax,
+            ImmutableArray<BoundExpression> elements, bool nullable, CSharpCompilation compilation, DiagnosticBag diagnostics, bool isRight)
+        {
+            ImmutableArray<TypeSymbol> convertedTypes = operators.SelectAsArray(o => o.GetConvertedType(isRight));
+            ImmutableArray<Location> elementLocations = elements.SelectAsArray(e => e.Syntax.Location);
+
+            var tuple = TupleTypeSymbol.Create(locationOpt: null, elementTypes: convertedTypes,
+                elementLocations, elementNames: default, compilation,
+                shouldCheckConstraints: true, errorPositions: default, syntax, diagnostics);
+
+            if (!nullable)
+            {
+                return tuple;
+            }
+
+            NamedTypeSymbol nullableT = GetSpecialType(SpecialType.System_Nullable_T, diagnostics, syntax);
+            NamedTypeSymbol nullableTuple = nullableT.Construct(tuple);
+            return nullableTuple;
+        }
+    }
+
+    /// <summary>
+    /// A tree of nested binary operators
+    /// </summary>
+    internal abstract class TupleBinaryOperatorInfo
```

Comment:
>TupleBinaryOperatorInfo [](start = 28, length = 23)

Can these types be defined in `BoundNodes.xml`?

Reply:
They could, but since they don't correspond to any specific syntax, I'm not sure it would be beneficial. I'm following the model used for foreach loops, awaits and deconstructions, which all use "info" objects.

---
In reply to: [153605157](https://github.com/dotnet/roslyn/pull/23356#discussion_r153605157) [](ancestors = 153605157)